security/rbac/enforcer.go

```diff
@@ -0,0 +1,154 @@
+package rbac
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/gaia-pipeline/gaia"
+)
+
+var (
+	errNamespaceNotFound = errors.New("namespace not found")
+	errActionNotFound    = errors.New("action not found")
+	errResourceNotFound  = errors.New("resource not found")
+	errResourceDeny      = errors.New("resource implicit deny")
+)
+
+// EnforcerConfig represents the config required for RBAC.
+type EnforcerConfig struct {
+	User      User
+	Namespace gaia.RBACPolicyNamespace
+	Action    gaia.RBACPolicyAction
+	Resource  gaia.RBACPolicyResource
+}
+
+// User represents the user to apply the enforcement to.
+type User struct {
+	Username string
+	Policies map[string]interface{}
+}
+
+// PolicyEnforcer is for enforcing RBAC Policies.
+type PolicyEnforcer interface {
+	Enforce(cfg EnforcerConfig) error
+	Evaluate(user User) (gaia.RBACEvaluatedPermissions, error)
+}
+
+type policyEnforcer struct {
+	svc Service
+}
+
+// NewPolicyEnforcer creates a new policyEnforcer
+func NewPolicyEnforcer(svc Service) PolicyEnforcer {
+	return &policyEnforcer{svc: svc}
+}
+
+// Enforce takes an EnforcerConfig containing User information (name, policies) and the required namespace, action and
+// resource to enforce. Evaluate all the users permissions into a more efficient single map based structure
+// gaia.RBACEvaluatedPermissions. Using the gaia.RBACEvaluatedPermissions we check if a user has the required
+// permissions.
+func (s *policyEnforcer) Enforce(cfg EnforcerConfig) error {
+	resolved, err := s.Evaluate(cfg.User)
+	if err != nil {
+		return fmt.Errorf("error evaluating policies: %v", err.Error())
+	}
+
+	ns, nsExists := resolved[cfg.Namespace]
+	if !nsExists {
+		return errNamespaceNotFound
+	}
+
+	act, actionExists := ns[cfg.Action]
+	if !actionExists {
+		return errActionNotFound
+	}
+
+	if cfg.Resource == "" {
```

Comment:
If the resource is empty it's all resources? I would think if the resource is empty, we shouldn't automatically allow all. `*` should be something very explicit. If new resources are defined, that means nothing is allowed.

Reply:
Yep, this was just temporary. I hadn’t got to implementing proper resource authentication yet mate 👍🏻 Was just to unblock testing on!